testcases/kernel/syscalls/recvmmsg/recvmmsg01.c

```diff
@@ -0,0 +1,2389 @@
+			}
+			nleft -= nrecv;
+			rmm += nrecv;
+		} while (nleft > 0);
+	}
+
+	thread_join(&smmc.smmc_thread_call);
+
+	int nsent = smmc.smmc_value;
+	assert((size_t) nsent == n);
+}
+
+STATIC void con_do_multi_send_recv(con_t *conp)
+{
+	assert(conp->con_call_n >= 1 && conp->con_call_n <= iovec_max());
+	size_t n = conp->con_call_n;
+	mmsghdr_t smm[n];
+	mmsghdr_t rmm[n];
+
+	/*
+	 * Gather the arguments from all the sendmsg_call_t and the
+	 * recvmsg_call_t into the smm[] and rmm[] vectors for the
+	 * multi-message send and receive syscalls.
+	 *
+	 * All the send and recv flags must be the same, there is
+	 * a single argument for the flags in sendmmsg(2) and in
+	 * recvmmsg(2), there is not an argument per message.
+	 */
+
+	int sflags = conp->con_smc_vec[0]->smc_flags;
+	int rflags = conp->con_rmc_vec[0]->rmc_flags;
+	size_t i;
+	for (i = 0; i < n; i++) {
+		assert(conp->con_smc_vec[i]->smc_flags == sflags);
+		assert(conp->con_rmc_vec[i]->rmc_flags == rflags);
+		mmsghdr_init(&smm[i], conp->con_smc_vec[i]->smc_msg);
+		mmsghdr_init(&rmm[i], conp->con_rmc_vec[i]->rmc_msg);
+	}
+
+	con_sendmmsg_recvmmsg(conp, n, sflags, smm, rflags, rmm);
+
+	/*
+	 * Scatter the results from smm[] and rmm[].
+	 */
+
+	for (i = 0; i < n; i++) {
+		conp->con_smc_vec[i]->smc_value = smm[i].msg_len;
+		conp->con_smc_vec[i]->smc_msg->msg_flags =
+						  smm[i].msg_hdr.msg_flags;
+		conp->con_rmc_vec[i]->rmc_value = rmm[i].msg_len;
+		conp->con_rmc_vec[i]->rmc_msg->msg_flags =
+						  rmm[i].msg_hdr.msg_flags;
+	}
+}
+
+STATIC void con_init_socketpair(con_t *conp)
+{
+	int pair[2];
+	SAFE_SOCKETPAIR(AF_UNIX, SOCK_SEQPACKET, 0, pair);
+	con_init(conp, pair[0], pair[1], SOCK_SEQPACKET, NULL);
+}
+
+STATIC void con_init_seqpacket(con_t *conp)
+{
+	con_init_with_type(conp, SOCK_SEQPACKET);
+}
+
+STATIC void con_init_dgram(con_t *conp)
+{
+	con_init_with_type(conp, SOCK_DGRAM);
+}
+
+STATIC void con_init_with_type(con_t *conp, int type)
+{
+#ifndef RECVMMSG_USE_SCTP
+	assert(type != SOCK_SEQPACKET);
+#endif
+	int protocol = (type == SOCK_SEQPACKET) ? IPPROTO_SCTP : 0;
+	int server = SAFE_SOCKET(PF_INET, type, protocol);
+
+	sockaddr_in_t sin;
+	sockaddr_in_init(&sin, AF_INET, TEST_PORT, TEST_IP);
+	SAFE_BIND(server, (sockaddr_t *) &sin, sizeof(sin));
+
+	int client = SAFE_SOCKET(PF_INET, type, protocol);
+
+	if (type == SOCK_SEQPACKET)
+		SAFE_LISTEN(server, 1);
+
+	con_init(conp, client, server, type, &sin);
+}
+
+STATIC void con_deinit(con_t *conp)
+{
+	close(conp->con_client);
+	close(conp->con_server);
+}
+
+STATIC void con_add_send_recv_calls(con_t *conp, sendmsg_call_t *smcp,
+				    recvmsg_call_t *rmcp, bool sender_sleeps)
+{
+	if (conp->con_call_n > 0) {
+		con_add_send_recv_calls_vec(conp, smcp, rmcp, sender_sleeps);
+		return;
+	}
+
+	thread_call(&smcp->smc_thread_call, (thread_start_t)
+		    (sender_sleeps ? sendmsg_call_after_sleep : sendmsg_call));
+	timeval_t pretv, postv;
+	if (sender_sleeps)
+		SAFE_GETTIMEOFDAY(&pretv, NULL);
```

Comment:
Using gettimeofday() for time measurement is broken by design. The wall clock could be changed by several different things on the system including ntpd daemon.

You have to use monotonic POSIX clock timer, see include/tst_timer.h, we have helper functions for time measurements there.